hazelcast/src/main/java/com/hazelcast/executor/impl/durable/DurableRingBuffer.java

```diff
@@ -0,0 +1,253 @@
+     */
+    public int add(Callable task) {
+        int index = findEmptySpot();
+        callableCounter++;
+        ringItems[index] = task;
+        isTask[index] = true;
+        sequences[index] = head;
+        return head;
+    }
+
+    private int findEmptySpot() {
+        if (callableCounter == ringItems.length) {
+            throw new RejectedExecutionException("Capacity[" + ringItems.length + "] is reached!!! ");
+        }
+        for (Object ringItem : ringItems) {
+            head++;
+            int index = toIndex(head);
+            if (!isTask[index]) {
+                return index;
+            }
+        }
+        throw new IllegalStateException();
+    }
+
+    /**
+     * Removes the task with the given sequence
+     *
+     * @param sequence the sequence
+     */
+    public void remove(int sequence) {
+        int index = toIndex(sequence);
+        ringItems[index] = null;
+        isTask[index] = false;
+        head--;
+        callableCounter--;
+    }
+
+    /**
+     * Puts the task for the given sequence
+     *
+     * @param sequence The sequence
+     * @param task     The task
+     */
+    void putBackup(int sequence, Callable task) {
+        head = Math.max(head, sequence);
+        callableCounter++;
+        int index = toIndex(sequence);
+        ringItems[index] = task;
+        isTask[index] = true;
+        sequences[index] = sequence;
+    }
+
+    /**
+     * Replaces the task with its response
+     * If the sequence does not correspond to a task then the call is ignored
+     *
+     * @param sequence The sequence
+     * @param response The response
+     */
+    void replaceTaskWithResult(int sequence, Object response) {
+        int index = toIndex(sequence);
+        // If sequence is not equal then it is disposed externally
+        if (sequences[index] != sequence) {
+            return;
+        }
+        ringItems[index] = response;
+        isTask[index] = false;
+        callableCounter--;
+    }
+
+    /**
+     * Gets the response and disposes the sequence
+     *
+     * @param sequence The sequence
+     * @return The response
+     */
+    Object retrieveAndDispose(int sequence) {
+        int index = toIndex(sequence);
+        checkSequence(index, sequence);
+        try {
+            return ringItems[index];
+        } finally {
+            ringItems[index] = null;
+            isTask[index] = false;
+            head--;
+        }
+    }
+
+    /**
+     * Disposes the sequence
+     *
+     * @param sequence The sequence
+     */
+    public void dispose(int sequence) {
+        int index = toIndex(sequence);
+        checkSequence(index, sequence);
+        if (isTask[index]) {
+            callableCounter--;
+        }
+        ringItems[index] = null;
+        isTask[index] = false;
+    }
+
+    /**
+     * Gets the response
+     *
+     * @param sequence The sequence
+     * @return The response
+     */
+    public Object retrieve(int sequence) {
```

Comment:
It is protected by WaitNotifyService. `DurableRetrieveResultOperation` runs only if `shouldWait` method returns false. `shouldWait` method checks if the result put to ringbuffer, returns true if it is not there yet.